src/sigma/r1_proof_generator.hpp

```diff
@@ -13,71 +13,90 @@ R1ProofGenerator<Exponent,GroupElement>::R1ProofGenerator(
 
 template<class Exponent, class GroupElement>
 void R1ProofGenerator<Exponent,GroupElement>::proof(
-        std::vector<Exponent>& a,
-        R1Proof<Exponent, GroupElement>& proof_out) const {
-    Exponent rA, rC, rD;
-    rA.randomize();
-    rC.randomize();
-    rD.randomize();
-    a.resize(n_ * m_);
-    for(int j = 0; j < m_; ++j){
-        for(int i = 1; i < n_; ++i){
-            a[j * n_ + i].randomize();
-            a[j * n_] -= a[j * n_ + i];
+        std::vector<Exponent>& a_out,
+        R1Proof<Exponent, GroupElement>& proof_out,
+        bool skip_final_response) {
+    rA_.randomize();
+    rC_.randomize();
+    rD_.randomize();
+    a_out.resize(n_ * m_);
+    for(int j = 0; j < m_; ++j) {
+        for(int i = 1; i < n_; ++i) {
+            a_out[j * n_ + i].randomize();
+            a_out[j * n_] -= a_out[j * n_ + i];
         }
     }
+
+    // proof_out.B_ = B_Commit;
+
     //compute A
     GroupElement A;
-    SigmaPrimitives<Exponent, GroupElement>::commit(g_, h_, a, rA, A);
+    SigmaPrimitives<Exponent, GroupElement>::commit(g_, h_, a_out, rA_, A);
     proof_out.A_ = A;
+
     //compute C
     std::vector<Exponent> c;
     c.resize(n_ * m_);
     for(int i = 0; i < n_ * m_; ++i) {
-        c[i] = (a[i] * (Exponent(uint64_t(1)) - (Exponent(uint64_t(2)) * b_[i])));
+        c[i] = (a_out[i] * (Exponent(uint64_t(1)) - (Exponent(uint64_t(2)) * b_[i])));
     }
     GroupElement C;
-    SigmaPrimitives<Exponent, GroupElement>::commit(g_,h_, c, rC, C);
+    SigmaPrimitives<Exponent, GroupElement>::commit(g_, h_, c, rC_, C);
     proof_out.C_ = C;
+
     //compute D
     std::vector<Exponent> d;
     d.resize(n_ * m_);
     for(int i = 0; i < n_ * m_; i++) {
-        d[i] = ((a[i].square()).negate());
+        d[i] = ((a_out[i].square()).negate());
     }
     GroupElement D;
-    SigmaPrimitives<Exponent, GroupElement>::commit(g_,h_, d, rD, D);
-    Exponent x;
+    SigmaPrimitives<Exponent, GroupElement>::commit(g_,h_, d, rD_, D);
+
     proof_out.D_ = D;
-    SigmaPrimitives<Exponent, GroupElement>::get_x(A, C, D,x);
-    x_ = x;
+
+    if (!skip_final_response) {
+        Exponent x;
+        std::vector<GroupElement> group_elements = {A, B_Commit, C, D};
+        SigmaPrimitives<Exponent, GroupElement>::generate_challenge(group_elements, x);
+        generate_final_response(a_out, x, proof_out);
+    }
+}
+
+template<class Exponent, class GroupElement>
+void R1ProofGenerator<Exponent,GroupElement>::generate_final_response(
+        const std::vector<Exponent>& a,
+        const Exponent& challenge_x,
+        R1Proof<Exponent, GroupElement>& proof_out) {
     //f
-    std::vector<Exponent> f;
-    f.reserve(m_ * (n_ - 1));
-    for(int j = 0; j < m_; j++){
+    proof_out.f_.reserve(m_ * (n_ - 1));
+    for(int j = 0; j < m_; j++) {
         for(int i = 1; i < n_; i++)
-        f.emplace_back(b_[(j * n_) + i] * x + a[(j * n_) + i]);
+        proof_out.f_.emplace_back(b_[(j * n_) + i] * challenge_x + a[(j * n_) + i]);
```

Comment:
Same as previous issue. If it not intension clear `proof_out` before adding new one will be more safe.

Reply:
done